language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -47,63 +47,66 @@ optionsParserInfo = info (optionsParser <**> helper)
     <> progDesc "Generate TypeScript bindings from a DAR"
     )
 
+-- Each package has a unique ID and is perhaps ascribed a symbolic
+-- name (that is, if a given package has been determined to be the
+-- "main-package" of some DAR).
+type PackageMap = Map.Map PackageId (Maybe String, Package)
+
+-- Merge a DAR into a package map (two DARs may have one or more
+-- packages in common). Merging fails and an IO exception is raised
+-- if:
+--   * Two packages are ascribed the same symbolic name;
+--   * A single package is ascribed two different names.
+mergePackageMap :: PackageMap -> FilePath -> IO PackageMap
```

Comment:
You are still trying to do multiple things at once. As suggested before I would split this up:
1. First read each DAR independently. This requires IO but it should be just a `mapM` or something like that.
2. Then merge the info you got from reading the DARs and detect things like name collisions. This does not require IO.
3. Then either print the errors and exit or run the codegen.

Reply:
Ok